src/main/java/com/spotify/futures/CompletableFuturesExtra.java

```diff
@@ -68,4 +72,132 @@ private CompletableFuturesExtra() {
+          BiFunction<? super T, Throwable, ? extends CompletionStage<U>> fn) {
+    return dereference(future.handle(fn));
+  }
+
+  /**
+   * Returns a new CompletionStage that, when this stage completes
+   * exceptionally, is executed with this stage's exception as the
+   * argument to the supplied function.  Otherwise, if this stage
+   * completes normally, then the returned stage also completes
+   * normally with the same value.
+   *
+   * This differs from
+   * {@link java.util.concurrent.CompletionStage#exceptionally(java.util.function.Function)}
+   * in that the function should return a {@link java.util.concurrent.CompletionStage} rather than
+   * the value directly.
+   *
+   * @param fn the function to use to compute the value of the
+   * returned CompletionStage if this CompletionStage completed
+   * exceptionally
+   * @return the new CompletionStage
+   */
+  public static <T> CompletionStage<T> exceptionallyCompose(
+          CompletionStage<T> future,
+          Function<Throwable, ? extends CompletionStage<T>> fn) {
+    return dereference(wrap(future).exceptionally(fn));
+  }
+
+  /**
+   * This takes a stage of a stage of a value and returns a plain future of a value.
+   *
+   * @param future of a future of a value
+   * @return the completion stage of a value
+   */
+  public static <T> CompletionStage<T> dereference(
+          CompletionStage<? extends CompletionStage<T>> future) {
+    //noinspection unchecked
+    return future.thenCompose(Identity.INSTANCE);
+  }
+
+  /**
+   * check that a stage is completed.
+   * @param stage the stage.
+   * @throws IllegalStateException if the stage is not completed.
+   */
+  public static void checkCompleted(CompletionStage<?> stage) {
+    if (!stage.toCompletableFuture().isDone()) {
+      throw new IllegalStateException("future was not completed");
+    }
+  }
+
+  /**
+   * Get the value of a completed stage.
+   *
+   * @param stage a completed stage.
+   * @return the value of the stage if it has one.
+   * @throws IllegalStateException if the stage is not completed.
+   * @throws com.google.common.util.concurrent.UncheckedExecutionException
+   * if the future has failed with a non-runtime exception, otherwise
+   * the actual exception
+   */
+  public static <T> T getCompleted(CompletionStage<T> stage) {
+    CompletableFuture<T> future = stage.toCompletableFuture();
+    checkCompleted(future);
+    try {
+      return future.join();
+    } catch (CompletionException e) {
+      throw Throwables.propagate(e.getCause());
+    }
+  }
+
+  private enum Identity implements Function {
+    INSTANCE;
+
+    @Override
+    public Object apply(Object o) {
+      return o;
+    }
+  }
+
+  private enum WrapFunction implements Function {
+    INSTANCE;
+
+    @Override
+    public Object apply(Object o) {
+      return CompletableFuture.completedFuture(o);
+    }
+  }
+
+  private static <T> CompletionStage<CompletionStage<T>> wrap(CompletionStage<T> future) {
```

Comment:
If you would be picky, this method should come before the two enum definitions, because that is more common. The style guide we use internally says 'some logical order, which its maintainer could explain if asked'. (https://google.github.io/styleguide/javaguide.html#s3.4.2-class-member-ordering). That makes me think ```wrap()``` should be next to ```dereference``` because they are similar.